File: src/main/java/kimspring/splearn/domain/member/MemberDetail.java
package kimspring.splearn.domain.member;

import jakarta.persistence.Entity;
import kimspring.splearn.domain.AbstractEntity;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.time.LocalDateTime;
import java.util.Objects;

import static org.springframework.util.Assert.isTrue;

@Entity
@Getter
@ToString(callSuper = true)
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class MemberDetail extends AbstractEntity {
    private Profile profile;

    private String introduction;

    private LocalDateTime registeredAt;

    private LocalDateTime activatedAt;

    private LocalDateTime deactivatedAt;

    static MemberDetail create() {
        MemberDetail memberDetail = new MemberDetail();
        memberDetail.registeredAt = LocalDateTime.now();
        return memberDetail;
    }

    void activatedAt() {
        isTrue(activatedAt == null, "이미 activatedAt은 설정되었습니다");
        this.activatedAt = LocalDateTime.now();
    }

    void deactivatedAt() {
        isTrue(deactivatedAt == null, "이미 deactivatedAt은 설정되었습니다");
        this.deactivatedAt = LocalDateTime.now();
    }

    void updateInfo(MemberInfoUpdateRequest updateRequest) {
        this.profile = new Profile(updateRequest.profileAddress());
        this.introduction = Objects.requireNonNull(updateRequest.introduction());
    }
}

File: src/main/kotlin/kimspring/splearn/domain/member/MemberDetail.kt
package kimspring.splearn.domain.member

import java.time.LocalDateTime

data class MemberDetail(
    val id: Long? = null,
    val profile: Profile? = null,
    val introduction: String? = null,
    val registeredAt: LocalDateTime,
    val activatedAt: LocalDateTime? = null,
    val deactivatedAt: LocalDateTime? = null,
) {
    fun recordActivation(now: LocalDateTime): MemberDetail {
        check(activatedAt == null) { "이미 activatedAt은 설정되었습니다" }
        return copy(activatedAt = now)
    }

    fun recordDeactivation(now: LocalDateTime): MemberDetail {
        check(deactivatedAt == null) { "이미 deactivatedAt은 설정되었습니다" }
        return copy(deactivatedAt = now)
    }

    fun updateInfo(
        profileAddress: String,
        introduction: String,
    ): MemberDetail = copy(profile = Profile(profileAddress), introduction = introduction)

    companion object {
        fun create(now: LocalDateTime): MemberDetail = MemberDetail(registeredAt = now)
    }
}
